Add Meshes to the Video RAM Profiler

Ryan-000 opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

![image](https://github.com/user-attachments/assets/1c54a980-74d6-48f4-88f9-b3741ebdb872)

Ryan-000 committed:
Add Meshes to the Video RAM Profiler

Apply suggestions from code review

Co-Authored-By: Hugo Locurcio <hugo.locurcio@hugo.pro>

Calinou reviewed (commented):
Tested locally, it works as expected in Forward+ and Mobile.

Some feedback:

- Meshes are not listed when using Compatibility. This makes the total memory usage look lower than it really is compared to other rendering methods.
  - This should either be implemented, or a message should be added somewhere when using Compatibility to tell the user that meshes aren't tracked in the video RAM profiler.
- Since meshes now have an icon displayed next to them, textures should also have a texture icon to ensure alignment. I'd probably use `Texture2D` for this, but perhaps `Cubemap` and `CubemapArray` can be used when relevant.
- There are a lot of unnamed entries with few vertices – sometimes even 0 vertices, amounting to 0 bytes. Should we hide entries that amount to 0 bytes from the list?

![image](https://github.com/user-attachments/assets/f495d5bf-2f0a-49d9-883b-375dc47b54b8)

This also occasionally happens with textures, at least in Compatibility:

![image](https://github.com/user-attachments/assets/2f487423-9784-4e0f-91eb-927b78efcdd2)

Also note the bottommost entry that matches the viewport resolution and is listed at 0 bytes. Is this a reference to the swapchain being held? I'd expect it to take about 3 MB of VRAM (`1152*648*4`).

Ryan-000 commented:
>Meshes are not listed when using Compatibility. This makes the total memory usage look lower than it really is compared to other rendering methods. This should either be implemented, or a message should be added somewhere when using Compatibility to tell the user that meshes aren't tracked in the video RAM profiler.

I added a notice tooltip about meshes not being tracked in compatibility as well as buffer memory (in both)

>Since meshes now have an icon displayed next to them, textures should also have a texture icon to ensure alignment. I'd probably use Texture2D for this, but perhaps Cubemap and CubemapArray can be used when relevant.

There was a bug when porting the profiler from godot 3 to godot 4 (Texture was renamed to Texture2D), I updated the code so it now shows the proper resource icon.

>There are a lot of unnamed entries with few vertices – sometimes even 0 vertices, amounting to 0 bytes. Should we hide entries that amount to 0 bytes from the list?

 I noticed that as well, so I've made mesh entries with no path and no VRAM usage hidden.

>This also occasionally happens with textures, at least in Compatibility: Also note the bottommost entry that matches the viewport resolution and is listed at 0 bytes. Is this a reference to the swapchain being held? I'd expect it to take about 3 MB of VRAM (1152*648*4).

I’m not sure what’s going on there. I don’t use Compatibility mode and didn’t modify any of the existing texture-related code (at the time you tested this), so whatever issue this is, it’s not related to this PR.


![image](https://github.com/user-attachments/assets/821b1817-ad19-4df8-ad57-697996b44c12)
Heres what it looks like now

Calinou commented on editor/debugger/script_editor_debugger.cpp:
```diff
@@ -2069,11 +2080,31 @@ ScriptEditorDebugger::ScriptEditorDebugger() {
+		vmlb->set_theme_type_variation("HeaderSmall");
 		vmem_hb->add_child(vmlb);
+
+		{ // add notice icon
```
```suggestion
		{ // Add notice icon.
```

Calinou commented on editor/debugger/script_editor_debugger.cpp:
```diff
@@ -2069,11 +2080,31 @@ ScriptEditorDebugger::ScriptEditorDebugger() {
+			vmem_hb->add_child(vmem_notice_icon);
+		}
+
+		{ // add some space to move the rest of the controls to the right
```
```suggestion
		{ // Add some space to move the rest of the controls to the right.
```

Calinou commented on editor/debugger/script_editor_debugger.cpp:
```diff
@@ -2069,11 +2080,31 @@ ScriptEditorDebugger::ScriptEditorDebugger() {
+			vmem_notice_icon->set_stretch_mode(TextureRect::STRETCH_KEEP_CENTERED);
+			vmem_notice_icon->set_h_size_flags(SIZE_SHRINK_CENTER);
+			vmem_notice_icon->set_visible(true);
+			vmem_notice_icon->set_tooltip_text(TTR(R"(Notice(s):
+This tool only reports tracked memory allocations, therefore total VRAM usage is inaccurate.
+Instead, use the monitors tab to obtain more precise VRAM usage.
+
+- Buffer Memory (e.g., GPUParticles) is not tracked
+- Meshes are not tracked in the Compatibility Renderer)"));
```
```suggestion
			vmem_notice_icon->set_tooltip_text(TTR(R"(Notice:
This tool only reports memory allocations tracked by the engine.
Therefore, total VRAM usage is inaccurate compared to what the Monitors tab or external tools can report.
Instead, use the monitors tab to obtain more precise VRAM usage.

- Buffer Memory (e.g., GPUParticles) is not tracked.
- Meshes are not tracked in the Compatibility renderer.)"));
```

Calinou commented on servers/debugger/servers_debugger.cpp:
```diff
@@ -454,9 +472,61 @@ void ServersDebugger::_send_resource_usage() {
+	for (const RS::MeshInfo &E : mesh_info) {
+		ServersDebugger::ResourceInfo info;
+		info.path = E.path;
+		// we use 64bit integers to avoid overflow, if for whatever reason, the sum is bigger than 4GB
```
```suggestion
		// We use 64-bit integers to avoid overflow, if for whatever reason, the sum is bigger than 4GB.
```

Calinou commented on servers/debugger/servers_debugger.cpp:
```diff
@@ -454,9 +472,61 @@ void ServersDebugger::_send_resource_usage() {
+		info.path = E.path;
+		// we use 64bit integers to avoid overflow, if for whatever reason, the sum is bigger than 4GB
+		uint64_t vram = E.vertex_buffer_size + E.attribute_buffer_size + E.skin_buffer_size + E.index_buffer_size + E.blend_shape_buffer_size + E.lod_index_buffers_size;
+		// but can info.vram even hold that, and why is it an int instead of an uint?
```
```suggestion
		// But can info.vram even hold that, and why is it an int instead of an uint?
```

Calinou commented on servers/debugger/servers_debugger.cpp:
```diff
@@ -454,9 +472,61 @@ void ServersDebugger::_send_resource_usage() {
+		Ref<Resource> resource = ResourceCache::get_ref(p_path);
+		return resource->get_class();
+	} else {
+		// this doesn't work all the time for embedded resources
```
```suggestion
		// This doesn't work all the time for embedded resources.
```

Calinou commented on servers/debugger/servers_debugger.cpp:
```diff
@@ -454,9 +472,61 @@ void ServersDebugger::_send_resource_usage() {
 	EngineDebugger::get_singleton()->send_message("servers:memory_usage", usage.serialize());
 }
 
+// done on a best effort basis.
```
```suggestion
// Done on a best-effort basis.
```

Calinou commented on editor/debugger/script_editor_debugger.cpp:
```diff
@@ -2184,11 +2195,32 @@ ScriptEditorDebugger::ScriptEditorDebugger() {
-		vmlb->set_theme_type_variation("HeaderSmall");
 
-		vmlb->set_h_size_flags(SIZE_EXPAND_FILL);
+		Label *vmlb = memnew(Label(TTR("List of Video Memory Usage by Resource:")));
```
```suggestion
		Label *vmlb = memnew(Label(TTRC("List of Video Memory Usage by Resource:")));
```

@KoBeWi Is this correct?

KoBeWi replied:
Yes, the tooltip should also use TTRC.

Calinou reviewed (commented):
Tested again, it works as expected. UX is good to me, I only have some comments about localization.

Ryan-000 commented on editor/debugger/script_editor_debugger.cpp:
```diff
@@ -2184,11 +2195,32 @@ ScriptEditorDebugger::ScriptEditorDebugger() {
+Therefore, total VRAM usage is inaccurate compared to what the Monitors tab or external tools can report.
+Instead, use the monitors tab to obtain more precise VRAM usage.
+
+- Buffer Memory (e.g., GPUParticles) is not tracked.
```
There are 2 instances of `TTR(R"` in https://github.com/godotengine/godot/blob/931820d33cbc38328e4014d65dd62c359b5116fc/editor/editor_audio_buses.cpp#L1399
and https://github.com/godotengine/godot/blob/8fd71e6b17811cfdd7cc222c97eff4f81edcf4c5/editor/editor_audio_buses.cpp#L1380

But I can still change it to TTRC if needed

clayjohn commented on servers/debugger/servers_debugger.cpp:
```diff
@@ -444,9 +462,61 @@ void ServersDebugger::_send_resource_usage() {
 }
 
+// Done on a best-effort basis.
+String ServersDebugger::_get_resource_type_from_path(const String &p_path) {
```
This is a nice addition. It makes me realize that we may be able to link back to the path in the file system somehow so people can just click to open the resource instead of tracing the path

Ryan-000 replied:
I believe this was already implemented in https://github.com/godotengine/godot/pull/103949

clayjohn replied:
Oh! Very nice

clayjohn reviewed (approved):
Looks great!

Repiteo commented:
Thanks!
